solutions/platinum/usaco-1188.mdx

```diff
@@ -0,0 +1,141 @@
+
+Now we need to swap it back towards the right until it reaches a point where which makes the array lexographically minimum. We can do this by finding the first block to the right of it that has it's $MAX > a[i]$ by checking each block again in $\mathcal{O}(\sqrt{N})$.
+
+Notice that there is an edge case here where $a[i]$ can never be swapped down to the left past another element that is greater than it because of the restriction of $k$ even thought the $MAX$ element is greater than $a[i]$. In this case, we can push $a[i]$ into the next block and then apply the strategy of checking when the $MAX > a[i]$. We can check this by seeing that if after inserting $a[i]$ into the block and swapping towards the left and right appropriately as described above, $a[i]$ reaches the end of the block.
+
+Now the problem arises that when we add enough elements, the size of a block could potentially be bigger than $\sqrt{n}$. To deal with this issue we can simply split the block into two smaller blocks (each half the size of the original block) and insert them into the list of blocks. Inserting into a vector in C++ or ArrayList in Java, is dependent on the size of the array which in our case which takes $\mathcal{O}(\sqrt{N})$. We can prove that the amount of blocks is bounded to $2\sqrt{n}$.
+
+Processing each element of the array took $\mathcal{O}(\sqrt{N})$ time so the total time complexity is $\mathcal{O}(N\sqrt{N})$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\sqrt{N})$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int n, k;
+
+struct Block {
+	int hi, lo;
+	vector<int> vals;
+	Block() : hi(-INT32_MAX), lo(INT32_MAX) {}
+
+	Block(vector<int> a) {
+		hi = -INT32_MAX;
+		lo = INT32_MAX;
+		for (int i : a) {
+			vals.push_back(i);
+			hi = max(hi, i);
+			lo = min(lo, i);
+		}
+	}
+
+	// Adds element x to the block
+	void add(int x) {
+		lo = min(lo, x);
+		hi = max(hi, x);
+		vals.push_back(x);
+		int i = vals.size() - 1;
+		while (i > 0 && abs(vals[i] - vals[i - 1]) <= k) {
+			swap(vals[i], vals[i - 1]);
+			i--;
+		}
+		while (i < vals.size() - 1 && vals[i] > vals[i + 1]) {
+			swap(vals[i], vals[i + 1]);
+			i++;
+		}
+	}
+
+	/*
+	Checks whether the most optimal position for x is at the end,
+	if it is starts from the end.
+	*/
+	bool reaches_end(int x) {
+		vals.push_back(x);
+		int i = vals.size() - 1;
+		while (i > 0 && abs(vals[i] - vals[i - 1]) <= k) {
+			swap(vals[i], vals[i - 1]);
+			i--;
+		}
+		while (i < vals.size() - 1 && vals[i] > vals[i + 1]) {
+			swap(vals[i], vals[i + 1]);
+			i++;
+		}
+		bool ans = vals.back() == x;
+		vals.erase(begin(vals) + i);
+		return ans;
+	}
+};
+
+int main() {
+	cin >> n >> k;
+	int len = (int)floor(sqrt(n));
+	vector<Block> sqrt{Block()};
+
+	for (int i = 0; i < n; i++) {
+		int x;
+		cin >> x;
+		int j = sqrt.size() - 1;
+
+		// Finds the earliest block we can insert element x by swapping to the
+		// left
```

Comment:
can you make these two comments a bit more even
like move some words from the top line to the bottom